flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -441,7 +469,17 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
             // This might change if the log ownership changes (the method won't likely be needed).
             stateChangelogWriter.confirm(lastUploadedFrom, lastUploadedTo);
         }
-        keyedStateBackend.notifyCheckpointComplete(checkpointId);
+        Long materializationID = materializationIdByCheckpointId.remove(checkpointId);
+        if (materializationID != null) {
+            keyedStateBackend.notifyCheckpointComplete(materializationID);
+            pendingMaterializationConfirmations.remove(materializationID);
```

Comment:
I think we can introduce another `lastConfirmedMaterizationId` (simliar like `lastCheckpointId`), we only notifyCheckpoint complete to wrapped state backend if current materilzation id is larger than `lastConfirmedMaterizationId`. If so, we can also avoid to update `pendingMaterializationConfirmations` each time as we don't need to record checkpoints if this materilzation id is equal to `lastConfirmedMaterizationId`.

Reply:
As discussed offline, I'll use `pendingMaterializationConfirmations` to prevent double notifications.

Reply:
Implemented in ac5235e16ec43f58d558472e9f2bc6f8512336eb